### ISSUE TITLE
Calling to_datetime on a frozen ActiveSupport::TimeWithZone causes a RuntimeError

### ISSUE THREAD

**Issue by HashNotAdam** [rails/rails#28101]
### Steps to reproduce
Create a frozen instance of ActiveSupport::TimeWithZone and then call `to_datetime` on it. For example: `Time.zone.now.freeze.to_datetime`.

### Expected behavior
Returns an instance of DateTime with the timezone's UTC offset; this should not modify the TimeWithZone instance.

### Actual behavior
> RuntimeError: can't modify frozen ActiveSupport::TimeWithZone
> from /Users/arice/dev/rails/activesupport/lib/active_support/time_with_zone.rb:412:in `to_datetime'

This appears to be related to `8c402ef` which stores the value to an instance variable as a form of caching.

### System configuration
**Rails version**: 4.2.8

**Ruby version**: 2.4.0p0

**Comment by HashNotAdam** [comment]
It seems like the simplest solution would be to include `to_datetime` in the overload for `freeze`:

```ruby
def freeze
  period; utc; time # preload instance variables before freezing
  super
end
```

While expensive, I have to assume the thought is that people are generally freezing constants and so it is worth the cost.

Thoughts?

**Comment by Tonkpils** [comment]
This seems related to an issue I'm running into. Comparing `ActiveSupport::TimeWithZone` that are frozen raises this error is the idea that one should not be freezing these objects?

```
RuntimeError: can't modify frozen ActiveSupport::TimeWithZone

Stack trace:
./app/models/....:in `>=`
```

Why would sending `>=` attempt to modify the instance?

**Comment by HashNotAdam** [comment]
@Tonkpils This appears to be a different, yet related, issue.

Short version is that `>=` triggers a call of `to_time` in the `DateAndTime::Compatibility` module which implements memorization, similar to `to_datetime`. Since the memorization requires saving an instance variable, it is trying to modify the frozen instance.

This has already been fixed in the master branch in the same fashion as I fixed `to_datetime`, however, it is yet to be included in a version of Rails 4.2

In the master branch, the "offending" code has been moved into `ActiveSupport::TimeWithZone` and reads:
```
def to_time
  if preserve_timezone
    @to_time_with_instance_offset ||= getlocal(utc_offset)
  else
    @to_time_with_system_offset ||= getlocal
  end
end
```
